synnax/pkg/distribution/channel/verification/service.go

```diff
@@ -0,0 +1,215 @@
+}
+
+var (
+	_             config.Config[Config] = Config{}
+	DefaultConfig                       = Config{
+		ExpireWarning: 7 * 24 * time.Hour,
+	}
+)
+
+// Validate implements config.Properties.
+func (c Config) Validate() error {
+	v := validate.New("key")
+	validate.NotNil(v, "DB", c.DB)
+	validate.NonZero(v, "ExpireWarning", c.ExpireWarning)
+	return v.Error()
+}
+
+// Override implements config.Properties.
+func (c Config) Override(other Config) Config {
+	c.DB = override.Nil(c.DB, other.DB)
+	c.Ins = override.Zero(c.Ins, other.Ins)
+	c.ExpireWarning = override.If(c.ExpireWarning, other.ExpireWarning, other.ExpireWarning.Nanoseconds() != 0)
+	return c
+}
+
+type Service struct {
+	Config
+	// shutdown
+	shutdown io.Closer
+}
+
+// OpenService opens a new service by creating a key from the argument
+// key stored in DB.
+func OpenService(key string, cfgs ...Config) (*Service, error) {
+	cfg, err := config.New(DefaultConfig, cfgs...)
+	if err != nil {
+		return nil, err
+	}
+	service := &Service{Config: cfg}
+
+	// This stuff needs to be updated when I decide what to do with the
+	// background key validation process
+	sCtx, cancel := signal.Isolated(signal.WithInstrumentation(service.Ins))
+	service.shutdown = signal.NewShutdown(sCtx, cancel)
+
+	var ctx context.Context
+	if key == "" {
+		return service, nil
+	}
+
+	// user did enter a key
+	err = service.create(ctx, key)
+	if err != nil {
+		return service, err
+	}
+	sCtx.Go(service.updateKeyValidation)
+
+	return service, err
+}
+
+// Close will shutdown the service
+func (s *Service) Close() error {
+	return s.shutdown.Close()
+}
+
+// ValidateChannelCount returns an error if inUse is greater than the number of
+// channels the user is allowed to use.
+func (s *Service) ValidateChannelCount(ctx context.Context, inUse int64) error {
+	key, err := s.retrieve(ctx)
+
+	if err != nil {
+		if inUse > maxFreeChannels {
+			return errFreeLimit(maxFreeChannels)
+		}
+		return nil
+	}
+
+	if getExpirationDate(key).Before(time.Now()) {
+		if inUse > maxFreeChannels {
+			return errExpireKey(maxFreeChannels)
+		}
+		return nil
+	}
+
+	if channelsAllowed := getNumberOfChannels(key); inUse > channelsAllowed {
+		return errOverChannelLimit(int(channelsAllowed))
+	}
+
+	return nil
+}
+
+// IsExpired returns an error if the key stored in s is expired.
+func (s *Service) IsExpired(ctx context.Context) error {
+	key, err := s.retrieve(ctx)
+	if err != nil {
+		return errFreeLimit(maxFreeChannels)
+
+	}
+	if getExpirationDate(key).Before(time.Now()) {
+		return errExpireKey(maxFreeChannels)
+	}
+	return nil
+}
+
+func (s *Service) GetMaxFreeChannels() int {
+	return maxFreeChannels
+}
+
+// create sets  key in s.DB. Returns an error if the key is invalid - you can
+// only set a valid key in the database
+func (s *Service) create(ctx context.Context, key string) error {
+	err := validateKey(key)
+	if err != nil {
+		return err
+	}
+	return s.DB.Set(ctx, []byte("productKey"), []byte(key))

```

Comment:
All calls of `productKey` should be base64 encoded. Not to mention, they can be extracted outside of this function because they are constants.